lib/galaxy/authnz/keycloak_authnz.py

```diff
@@ -0,0 +1,147 @@
+                                                        id_token=id_token,
+                                                        refresh_token=refresh_token,
+                                                        expiration_time=expiration_time,
+                                                        refresh_expiration_time=refresh_expiration_time,
+                                                        raw_token=token)
+        else:
+            keycloak_access_token.access_token = access_token
+            keycloak_access_token.id_token = id_token
+            keycloak_access_token.refresh_token = refresh_token
+            keycloak_access_token.expiration_time = expiration_time
+            keycloak_access_token.refresh_expiration_time = refresh_expiration_time
+            keycloak_access_token.raw_token = token
+        trans.sa_session.add(keycloak_access_token)
+        trans.sa_session.flush()
+        return login_redirect_url, keycloak_access_token.user
+
+    def disconnect(self, provider, trans, disconnect_redirect_url=None, association_id=None):
+        # TODO: implement?
+        pass
+
+    def _create_oauth2_session(self, state=None):
+        client_id = self.config['client_id']
+        redirect_uri = self.config['redirect_uri']
+        return OAuth2Session(client_id,
+                             scope=('openid', 'email', 'profile'),
+                             redirect_uri=redirect_uri,
+                             state=state)
+
+    def _fetch_token(self, oauth2_session, trans):
+        client_secret = self.config['client_secret']
+        token_endpoint = self.config['token_endpoint']
+        return oauth2_session.fetch_token(
+            token_endpoint,
+            client_secret=client_secret,
+            authorization_response=trans.request.url)
+
+    def _get_userinfo(self, oauth2_session):
+        userinfo_endpoint = self.config['userinfo_endpoint']
+        return oauth2_session.get(userinfo_endpoint).json()
+
+    def _get_keycloak_access_token(self, sa_session, keycloak_user_id):
+        return sa_session.query(KeycloakAccessToken).filter_by(
+            keycloak_user_id=keycloak_user_id).one_or_none()
+
+    def _create_user(self, sa_session, username, email):
+        user = User(email=email, username=username)
+        user.set_random_password()
+        sa_session.add(user)
+        sa_session.flush()
+        return user
+
+    def _hash_nonce(self, nonce):
+        return hashlib.sha256(nonce).hexdigest()
+
+    def _validate_nonce(self, trans, nonce_hash):
+        nonce_cookie = trans.get_cookie(name=NONCE_COOKIE_NAME)
+        # Delete the nonce cookie
+        trans.set_cookie('', name=NONCE_COOKIE_NAME, age=-1)
+        nonce_cookie_hash = self._hash_nonce(nonce_cookie)
+        if nonce_hash != nonce_cookie_hash:
+            raise Exception("Nonce mismatch!")
+
+    def _load_well_known_oidc_config(self, well_known_uri):
+        return requests.get(well_known_uri).json()
```

Comment:
The URL should be "http" not "https".  See if that works.

Reply:
Thats correct; however, it might be better to handle such an exception here, and also update the example well_known url.

Reply:
Also with http I am getting `null` redirect uri, and checking terminal I see the following error:

```shell
  File "lib/galaxy/authnz/managers.py", line 233, in authenticate
    return True, "Redirecting to the `{}` identity provider for authentication".format(provider), backend.authenticate(trans)
  File "lib/galaxy/authnz/keycloak_authnz.py", line 44, in authenticate
    base_authorize_url, **extra_params)
  File "../.venv/lib/python2.7/site-packages/requests_oauthlib/oauth2_session.py", line 153, in authorization_url
    **kwargs), state
  File "../.venv/lib/python2.7/site-packages/oauthlib/oauth2/rfc6749/clients/web_application.py", line 86, in prepare_request_uri
    redirect_uri=redirect_uri, scope=scope, state=state, **kwargs)
  File "../.venv/lib/python2.7/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 71, in prepare_grant_uri
    raise InsecureTransportError()
InsecureTransportError: (insecure_transport) OAuth 2 MUST utilize https.
```
While this error makes a total sense, as it is per explicit requirements of OAuth 2.0 (and OIDC inherently) to TLS/SSL secure connection, but it can be tricky for local testing. Maybe this would better bond with the `VERIFY_SSL` setter `oidc_config.xml`. Those configurations should apply to all providers. So, while `VERIFY_SSL` is set to `False`, the backend should not throw exception/warning complaining of insecure connection.

Reply:
Did you create the environment variable: `export OAUTHLIB_INSECURE_TRANSPORT=1` ?